core/src/main/java/io/hyperfoil/core/handlers/json/StreamQueue.java

```diff
@@ -13,25 +14,40 @@
 public class StreamQueue {
    protected static final Logger log = LogManager.getLogger(StreamQueue.class);
 
-   private final ByteStream[] parts;
+   private ByteStream[] parts;
    // indices used by users for the beginning (readerIndex) of the bytestream
-   private final int[] userIndex;
-   private final int mask;
+   private int[] userIndex;
+   private int mask;
    private int end = -1;
    private int tail = 0;
    private int length = 0;
 
-   public StreamQueue(int capacity) {
-      this.mask = (1 << 32 - Integer.numberOfLeadingZeros(capacity - 1)) - 1;
-      this.parts = new ByteStream[capacity];
-      this.userIndex = new int[capacity];
+   public StreamQueue(int initialCapacity) {
+      initialCapacity = Pow2.roundToPowerOfTwo(initialCapacity);
+      this.mask = initialCapacity - 1;
+      this.parts = new ByteStream[initialCapacity];
+      this.userIndex = new int[initialCapacity];
       Arrays.fill(userIndex, -1);
    }
 
+   public void enlargeCapacity() {
+      assert parts[tail] != null;
+      int newCapacity = Pow2.roundToPowerOfTwo(mask + 2);
+      ByteStream[] newParts = Arrays.copyOf(parts, newCapacity);
+      int[] newUserIndex = Arrays.copyOf(userIndex, newCapacity);
+      Arrays.fill(newUserIndex, mask + 1, newCapacity, -1);
+      // help GC
+      Arrays.fill(parts, null);
+      mask = newCapacity - 1;
+      tail = end + 1;
+      userIndex = newUserIndex;
+      parts = newParts;
+      assert parts[tail] == null;
+   }
+
    public int append(ByteStream stream) {
       if (parts[tail] != null) {
-         log.warn("Too many buffered fragments, dropping data.");
-         parts[tail].release();
+         enlargeCapacity();
```

Comment:
I Will probably retain the warn here so we can spot that a top big number of fragments are received